py3dtiles/gltf.py

```diff
@@ -55,19 +55,21 @@ def from_array(array):
         if struct.unpack("4s", array[0:4])[0] != b"glTF":
             raise RuntimeError("Array does not contain a binary glTF")
 
-        if struct.unpack("i", array[4:8])[0] != 1:
+        version = struct.unpack("i", array[4:8])[0]
+        if version != 1 and version != 2:
             raise RuntimeError("Unsupported glTF version")
 
         length = struct.unpack("i", array[8:12])[0]
         content_length = struct.unpack("i", array[12:16])[0]
 
-        if struct.unpack("i", array[16:20])[0] != 0:
+        content_type = struct.unpack("i", array[16:20])[0]
+        if content_type != 0 and content_type != 1313821514:  # 1313821514 => JSON
             raise RuntimeError("Unsupported binary glTF content type")
 
         header = struct.unpack(str(content_length) + "s",
                                array[20:20 + content_length])[0]
         glTF.header = json.loads(header.decode("ascii"))
-        glTF.body = array[20 + content_length:length]
+        glTF.body = array[28 + content_length:length]
```

Comment:
Maybe something like '20 + content_length' would be easier to read than 28 ?
Ditto for all above constants like '20' that seems to be '16 + content_type_length'.
Intermediate variables are not that expensive for runtime but make things much easier to read.

Reply:
The '20' correspond to the size of the [binaryHeader](https://github.com/VCityTeam/py3dtiles/blob/Tiler/py3dtiles/gltf.py#L34) (12 bits -> 3 elements encoded as uint8) + the size of the [jsonChunkHeader](https://github.com/VCityTeam/py3dtiles/blob/Tiler/py3dtiles/gltf.py#L35) (8 bits -> 2 elements encoded as uint8). The 'content_length' is the size of the [scene](https://github.com/VCityTeam/py3dtiles/blob/Tiler/py3dtiles/gltf.py#L36). To create the header of the GlTF instance, we skip the 20 first bytes (binaryHeader and jsonChunkHeader) to keep only the content (`array[20 : 20 + content_length]`). Since the 20 first bytes will be added by the [to_array](https://github.com/VCityTeam/py3dtiles/blob/Tiler/py3dtiles/gltf.py#L14), we don't need them into the GlTF instance created by [from_array](https://github.com/VCityTeam/py3dtiles/blob/Tiler/py3dtiles/gltf.py#L42).

The '28' correspond the '20' explained above + the size of the [binChunkHeader](https://github.com/VCityTeam/py3dtiles/blob/Tiler/py3dtiles/gltf.py#L37) (8 bits -> 2 elements encoded as uint8).  
We used to write the content of the body starting at the index `20 + content_length`. But by doing that, we write the [binChunkHeader](https://github.com/VCityTeam/py3dtiles/blob/Tiler/py3dtiles/gltf.py#L30) into the body, then we add it again during the [to_array](https://github.com/VCityTeam/py3dtiles/blob/Tiler/py3dtiles/gltf.py#L14). To avoid having this binChunkHeader written twice (creating an error when reading the produced glTF), I corrected the starting index `20 + content_length` to `28 + content_length`.

Those '20' (12 + 8) and '28' (12 + 8 + 8) depend on the sizes of the [binaryHeader](https://github.com/VCityTeam/py3dtiles/blob/Tiler/py3dtiles/gltf.py#L24), [jsonChunkHeader](https://github.com/VCityTeam/py3dtiles/blob/Tiler/py3dtiles/gltf.py#L27) and [binChunkHeader](https://github.com/VCityTeam/py3dtiles/blob/Tiler/py3dtiles/gltf.py#L30). The [from_array](https://github.com/VCityTeam/py3dtiles/blob/Tiler/py3dtiles/gltf.py#L42) don't know those sizes, since the 3 headers are created only in the [to_array](https://github.com/VCityTeam/py3dtiles/blob/Tiler/py3dtiles/gltf.py#L14). All the [`struct.unpack...`](https://github.com/VCityTeam/py3dtiles/blob/Tiler/py3dtiles/gltf.py#L55) are also hardcoded to match those 3 headers. I don't know if there is a good way to read the glTF in the `from_array` method without hardcoding the offsets.

Reply:
![image](https://user-images.githubusercontent.com/32875283/148960609-24bc0196-8790-4b4d-8345-d0ff214a0609.png)  
With [glTF 2.0 specification](https://www.khronos.org/registry/glTF/specs/2.0/glTF-2.0.html#binary-gltf-layout), we see that the binary of the glTF contains:
- A 12-bytes header ([binaryHeader](https://github.com/VCityTeam/py3dtiles/blob/Tiler/py3dtiles/gltf.py#L24) in the code) with three 4-bytes entries
- A mandatory JSON chunk and an optionnal binary chunk:
    - Each junk has a header of two 4-bytes entries (length and type of the chunk)
        - [jsonChunkHeader](https://github.com/VCityTeam/py3dtiles/blob/Tiler/py3dtiles/gltf.py#L27) is the header of the JSON chunk in code
        -  [binChunkHeader](https://github.com/VCityTeam/py3dtiles/blob/Tiler/py3dtiles/gltf.py#L30) is the header of the binary chunk in code
    - Each junk has data
        -  the JSON [scene](https://github.com/VCityTeam/py3dtiles/blob/Tiler/py3dtiles/gltf.py#L15) for the JSON chunk
        -  the content of the scene (positions, normals, etc) for the binary chunk

Since the header must be 12-bytes long and each chunk header must be 8-bytes long, we could replace the hardcoded '20' and '28' by something like `GlTF.HEADER_LENGTH + GlTF.CHUNK_HEADER_LENGTH` and `GlTF.HEADER_LENGTH + (2 * GlTF.CHUNK_HEADER_LENGTH)`

Reply:
Yep, this looks much better for the next code reader !

Reply:
Variable name semantics is of real help (to me at least)